File: network/src/main/java/io/appmetrica/analytics/network/internal/Response.java
package io.appmetrica.analytics.network.internal;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.appmetrica.analytics.network.impl.utils.Utils;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Response {

    private final boolean completed;
    private final int code;
    @NonNull
    private final byte[] responseData;
    @NonNull
    private final byte[] errorData;
    @NonNull
    private final Map<String, List<String>> headers;
    @Nullable
    private final Throwable exception;

    public Response(@Nullable Throwable exception) {
        this(false, 0, new byte[0], new byte[0], new HashMap<String, List<String>>(), exception);
    }

    public Response(boolean completed,
                    int code,
                    @NonNull byte[] responseData,
                    @NonNull byte[] errorData,
                    @Nullable Map<String, List<String>> headers,
                    @Nullable Throwable exception) {
        this.completed = completed;
        this.code = code;
        this.responseData = responseData;
        this.errorData = errorData;
        this.headers = headers == null ? Collections.<String, List<String>>emptyMap() :
                Utils.unmodifiableMapCopy(headers);
        this.exception = exception;
    }

    public int getCode() {
        return code;
    }

    @NonNull
    public byte[] getResponseData() {
        return responseData;
    }

    @NonNull
    public byte[] getErrorData() {
        return errorData;
    }

    @NonNull
    public Map<String, List<String>> getHeaders() {
        return headers;
    }

    @Nullable
    public Throwable getException() {
        return exception;
    }

    public boolean isCompleted() {
        return completed;
    }

    @Override
    public String toString() {
        return "Response{" +
                "completed=" + completed +
                ", code=" + code +
                ", responseDataLength=" + responseData.length +
                ", errorDataLength=" + errorData.length +
                ", headers=" + headers +
                ", exception=" + exception +
                '}';
    }
}

File: network-api/src/main/java/io/appmetrica/analytics/networkapi/Response.kt
package io.appmetrica.analytics.networkapi

/**
 * Represents an HTTP response.
 * Use [Builder] to construct instances of this class.
 *
 * @property isCompleted Whether the request completed successfully without exceptions
 * @property code HTTP response code (e.g., 200, 404, 500)
 * @property responseData Response body as byte array
 * @property headers HTTP response headers as key-value pairs with multiple values per key
 * @property exception Exception that occurred during request execution, or null if successful
 * @property url Final URL after any redirects, or null if not available
 */
class Response private constructor(
    val isCompleted: Boolean,
    val code: Int,
    val responseData: ByteArray,
    val headers: Map<String, List<String>>,
    val exception: Throwable?,
    val url: String?,
) {

    override fun toString(): String {
        return "Response(" +
            "isCompleted=$isCompleted, " +
            "code=$code, " +
            "responseDataLength=${responseData.size}, " +
            "headers=$headers, " +
            "exception=$exception, " +
            "url=$url" +
            ")"
    }

    /**
     * Builder for creating [Response] instances.
     * Provides constructors for both successful responses and error responses.
     */
    class Builder private constructor(
        private val isCompleted: Boolean,
        private val code: Int,
        private val responseData: ByteArray,
        private val exception: Throwable?,
    ) {

        private var headers: Map<String, List<String>> = emptyMap()
        private var url: String? = null

        /**
         * Creates a builder for an error response with an exception.
         *
         * @param exception Exception that occurred during request execution
         */
        constructor(
            exception: Throwable?
        ) : this(
            isCompleted = false,
            code = 0,
            responseData = ByteArray(0),
            exception = exception,
        )

        /**
         * Creates a builder for a response with status code and data.
         *
         * @param isCompleted Whether the request completed successfully
         * @param code HTTP response code
         * @param responseData Response body as byte array
         */
        constructor(
            isCompleted: Boolean,
            code: Int,
            responseData: ByteArray,
        ) : this(
            isCompleted = isCompleted,
            code = code,
            responseData = responseData,
            exception = null,
        )

        /**
         * Sets the response headers.
         *
         * @param headers HTTP response headers as key-value pairs with multiple values per key
         * @return This builder instance for method chaining
         */
        fun withHeaders(headers: Map<String, List<String>>): Builder {
            this.headers = headers.toMap()
            return this
        }

        /**
         * Sets the final URL after any redirects.
         *
         * @param url Final URL
         * @return This builder instance for method chaining
         */
        fun withUrl(url: String): Builder {
            this.url = url
            return this
        }

        /**
         * Builds a new [Response] instance with the configured parameters.
         *
         * @return A new HTTP response instance
         */
        fun build(): Response {
            return Response(
                isCompleted = isCompleted,
                code = code,
                responseData = responseData,
                headers = headers,
                exception = exception,
                url = url,
            )
        }
    }
}
